File: frontend/src/api/order.ts
import { get, post } from '../utils/request';
import type { PageData } from './product';

export interface OrderItemVO {
  id: number;
  product_id: number;
  product_name: string;
  product_image: string;
  price: number;
  quantity: number;
  total_price: number;
  reviewed: boolean;
}

export interface OrderVO {
  id: number;
  order_no: string;
  user_id: number;
  status: string;
  status_text: string;
  total_amount: number;
  discount_amount: number;
  shipping_fee: number;
  pay_amount: number;
  coupon_id: number;
  receiver_name: string;
  receiver_phone: string;
  receiver_addr: string;
  remark: string;
  paid_at: string;
  shipped_at: string;
  completed_at: string;
  cancelled_at: string;
  created_at: string;
  items: OrderItemVO[];
}

export interface LogisticsEventVO {
  time: string;
  status: string;
  desc: string;
}

export interface LogisticsVO {
  tracking_no: string;
  carrier: string;
  status: string;
  status_text: string;
  events: LogisticsEventVO[];
}

export interface CheckoutRequest {
  cart_item_ids: number[];
  coupon_id?: number;
  receiver_name: string;
  receiver_phone: string;
  receiver_addr: string;
  remark?: string;
}

export const checkout = (data: CheckoutRequest) => post<{ message: string; order: OrderVO }>('/orders', data);

export const listOrders = (params?: { page?: number; page_size?: number; status?: string; order_no?: string }) =>
  get<PageData<OrderVO>>('/orders', params as Record<string, unknown>);

export const getOrder = (id: number) => get<OrderVO>(`/orders/${id}`);

export const payOrder = (id: number, channel = 'ALIPAY_SANDBOX') => post<{ message: string; order: OrderVO }>(`/orders/${id}/pay`, { channel });

export const cancelOrder = (id: number, reason?: string) => post<{ message: string; order: OrderVO }>(`/orders/${id}/cancel`, { reason });

export const completeOrder = (id: number) => post<{ message: string; order: OrderVO }>(`/orders/${id}/complete`);

export const getLogistics = (id: number) => get<LogisticsVO>(`/orders/${id}/logistics`);

File: backend/internal/model/order.go
package model

import "time"

// Order 订单实体：核心状态机 PENDING_PAYMENT → PENDING_SHIPMENT → SHIPPED → COMPLETED / CANCELLED。
type Order struct {
	ID             uint       `gorm:"primaryKey" json:"id"`
	OrderNo        string     `gorm:"size:32;uniqueIndex;not null" json:"order_no"`
	UserID         uint       `gorm:"index;not null" json:"user_id"`
	Status         string     `gorm:"size:24;default:PENDING_PAYMENT;index" json:"status"`
	TotalAmount    float64    `gorm:"type:decimal(10,2);not null" json:"total_amount"`
	DiscountAmount float64    `gorm:"type:decimal(10,2);default:0" json:"discount_amount"`
	ShippingFee    float64    `gorm:"type:decimal(10,2);default:0" json:"shipping_fee"`
	PayAmount      float64    `gorm:"type:decimal(10,2);not null" json:"pay_amount"`
	CouponID       uint       `gorm:"default:0" json:"coupon_id"`
	ReceiverName   string     `gorm:"size:64;not null" json:"receiver_name"`
	ReceiverPhone  string     `gorm:"size:32;not null" json:"receiver_phone"`
	ReceiverAddr   string     `gorm:"size:255;not null" json:"receiver_addr"`
	Remark         string     `gorm:"size:255" json:"remark"`
	PaidAt         *time.Time `json:"paid_at"`
	ShippedAt      *time.Time `json:"shipped_at"`
	CompletedAt    *time.Time `json:"completed_at"`
	CancelledAt    *time.Time `json:"cancelled_at"`
	CancelledBy    string     `gorm:"size:32" json:"cancelled_by"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`

	Items []OrderItem `gorm:"foreignKey:OrderID" json:"items,omitempty"`
}
